verifiable-api/server/bin/handlers.rs

```diff
@@ -0,0 +1,439 @@
+            ));
+        }
+
+        let mut filter = Filter::new();
+        if let Some(from_block) = self.from_block {
+            filter = filter.from_block(from_block);
+        }
+        if let Some(to_block) = self.to_block {
+            filter = filter.to_block(to_block);
+        }
+        if let Some(block_hash) = self.block_hash {
+            filter = filter.at_block_hash(block_hash);
+        }
+        if !self.address.is_empty() {
+            filter = filter.address(self.address);
+        }
+        if !self.topic0.is_empty() {
+            filter = filter.event_signature(FilterSet::from(self.topic0));
+        }
+        if !self.topic1.is_empty() {
+            filter = filter.topic1(FilterSet::from(self.topic1));
+        }
+        if !self.topic2.is_empty() {
+            filter = filter.topic2(FilterSet::from(self.topic2));
+        }
+        if !self.topic3.is_empty() {
+            filter = filter.topic3(FilterSet::from(self.topic3));
+        }
+
+        Ok(filter)
+    }
+}
+
+/// This method returns the EIP-1186 account proof for a given address.
+///
+/// Replaces the `eth_getProof`, `eth_getTransactionCount`,
+/// `eth_getBalance` `eth_getCode` and `eth_getStorageAt`RPC methods.
+pub async fn get_account<N: NetworkSpec, R: ExecutionRpc<N>>(
+    Path(address): Path<Address>,
+    Query(AccountProofQuery {
+        storage_slots,
+        block,
+    }): Query<AccountProofQuery>,
+    State(ApiState { rpc, .. }): State<ApiState<N, R>>,
+) -> Response<AccountResponse> {
+    // Ensure that BlockId is of block number variant
+    let block = block.unwrap_or(BlockId::latest());
+    let block_num = match block {
+        BlockId::Number(number_or_tag) => match number_or_tag {
+            BlockNumberOrTag::Number(number) => Ok(number),
+            tag => rpc
+                .get_block_by_number(tag, BlockTransactionsKind::Hashes)
+                .await
+                .map_err(map_server_err)?
+                .ok_or_else(|| (StatusCode::BAD_REQUEST, json_err("Block not found")))
+                .map(|block| block.header().number()),
+        },
+        BlockId::Hash(hash) => rpc
+            .get_block(hash.into())
+            .await
+            .map_err(map_server_err)
+            .map(|block| block.header().number()),
+    }?;
+    let block = BlockId::from(block_num);
+
+    let storage_keys = storage_slots
+        .into_iter()
+        .map(|key| key.into())
+        .collect::<Vec<_>>();
+    let proof = rpc
+        .get_proof(address, &storage_keys, block)
+        .await
+        .map_err(map_server_err)?;
+
+    let code = rpc
+        .get_code(address, block_num)
+        .await
+        .map_err(map_server_err)?;
+
+    Ok(Json(AccountResponse {
+        account: Account {
+            balance: proof.balance,
+            nonce: proof.nonce,
+            code_hash: proof.code_hash,
+            storage_root: proof.storage_hash,
+        },
+        code: code.into(),
```

Comment:
What if we don't care about the code (for example, in the case we are fetching a balance). We should be able to replace it with a codehash in those instances.

Reply:
So my idea was to basically replace the `eth_getProof`, `eth_getTransactionCount`, `eth_getBalance` `eth_getCode` and `eth_getStorageAt` methods with a single endpoint that is `/account/<address>` here.

Currently in Helios, methods like `eth_getBalance`, `eth_getTransactionCount`, `eth_getStorageAt` also fetch `eth_getProof` and `eth_getCode` to verify the account proof.

A single API that returns all required account related data makes it more clean IMHO.

Reply:
It is certainly more clean, but my concern is that contract code can be up to 25kb, which is not an insubstantial amount to waste if we don't need it.

We could have a query parameter that decides whether to include the full code and make the code parameter an optional. We should be sure to document that None does not necessarily mean there is no code at the location, but that we didn't ask for it.

Reply:
Ah we probably shouldn't also fetch the code when just verifying a balance. The codehash is always correct since it is hashed into the account, which is verified via merkle proof against the stateroot. So this means we don't actually need to fetch the code if we do not care about it.

Reply:
> We should be sure to document that None does not necessarily mean there is no code at the location, but that we didn't ask for it.

Could have a query param `?includeCode=<bool>` that omits the `code` field from response if `false`. Don't need to handle the `None` case separately then.

Reply:
Yep I think thats fine. For deserialization purposed you will probably have an easier time having it be an optional in the struct, which when the json omits it deserializes to none.

Reply:
Made the `code` field optional in 77dbd2809f29c638cd3068f09c75e9c3b8eaa604.